pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java

```diff
@@ -99,28 +116,43 @@ public int submit(long requestId, QueryPlan queryPlan, long timeoutMs, Map<Strin
           String host = serverInstance.getHostname();
           int servicePort = serverInstance.getQueryServicePort();
           DispatchClient client = getOrCreateDispatchClient(host, servicePort);
-          Worker.QueryResponse response = client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
-                  QueryPlanSerDeUtils.serialize(constructDistributedStagePlan(queryPlan, stageId, serverInstance)))
-              .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
-              .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs))
-              .putAllMetadata(queryOptions).build());
-
-          if (response.containsMetadata(QueryConfig.KEY_OF_SERVER_RESPONSE_STATUS_ERROR)) {
-            throw new RuntimeException(
-                String.format("Unable to execute query plan at stage %s on server %s: ERROR: %s", stageId,
-                    serverInstance, response));
+          dispatchCalls++;
+          _executorService.submit(() -> {
+            client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
+                QueryPlanSerDeUtils.serialize(constructDistributedStagePlan(queryPlan, stageId, serverInstance)))
+                .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
+                .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs))
+                .putAllMetadata(queryOptions).build(), stageId, serverInstance, deadline, callbacks::offer);
+          });
+        }
+      }
+    }
+    for (int i = 0; i < dispatchCalls; i++) {
+      AsyncResponse resp;
+      while (!deadline.isExpired()) {
+        resp = callbacks.poll(DEFAULT_DISPATCHER_CALLBACK_POLL_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        if (resp != null) {
+          if (resp.getThrowable() != null) {
+            throw new RuntimeException(String.format("Error dispatching query to server=%s stage=%s",
+                resp._virtualServer, resp._stageId), resp.getThrowable());
+          } else {
+            Worker.QueryResponse response = resp.getQueryResponse();
+            if (response.containsMetadata(QueryConfig.KEY_OF_SERVER_RESPONSE_STATUS_ERROR)) {
+              throw new RuntimeException(
+                  String.format("Unable to execute query plan at stage %s on server %s: ERROR: %s", resp.getStageId(),
+                      resp.getVirtualServer(), response));
+            }
+            break;
```

Comment:
not sure why we need a separate deadline expire + 2 nested loop. 
cant we just do
```
while (!deadline.isExpired() && successfulDispatchCount < dispatchCalls) {
  resp = callbacks.poll(TIMEOUT);
  if (...) {
     .... // all the error conditions.
  } else {
    successfulDispatch ++;
  }
}
```